Completing ring_series

shivamvats opened the pull request:
This PR ports part of #609. I have not yet added tests, but speed comparisons look very promising.

```
In [27]: %timeit t = rs_atan(x,x,500)
10 loops, best of 3: 41.4 ms per loop

In [28]: %timeit s= atan(z).series(z,0,500)
1 loops, best of 3: 9.89 s per loop

In [29]: %timeit t = rs_sin(x, x, 500)
100 loops, best of 3: 13.7 ms per loop

In [30]: %timeit s = sin(z).series(z, 0, 500)
1 loops, best of 3: 10 s per loop

In [17]: %timeit t = rs_tan(x, x, 100)
1 loops, best of 3: 178 ms per loop

In [18]: %timeit s = tan(z).series(z, 0, 100)
1 loops, best of 3: 1.68 s per loop
```

@certik @smichr @pernici Please review.

shivamvats committed:
Ported atan to ring_series as rs_atan

shivamvats committed:
Added fun function

shivamvats committed:
Ported sin as rs_sin

shivamvats committed:
Made documentation fixes

shivamvats committed:
Fixed multivariate series expansion for rs_tan

shivamvats committed:
Typos fixed

certik commented on sympy/polys/ring_series.py:
```diff
@@ -419,6 +418,61 @@ def rs_integrate(self, x):
+    ========
+
+    >>> from sympy.polys.domains import QQ
+    >>> from sympy.polys.rings import ring
```
You need to import `fun` here as well. See the Travis [failure](https://travis-ci.org/sympy/sympy/jobs/57463965):
```
_________________________ sympy.polys.ring_series.fun __________________________

File "/home/travis/virtualenv/python2.6.9/lib/python2.6/site-packages/sympy-0.7.6_git-py2.6.egg/sympy/polys/ring_series.py", line 448, in sympy.polys.ring_series.fun

Failed example:

fun(p, _tan1, 'x', 4)

Exception raised:

Traceback (most recent call last):

File "/opt/python/2.6.9/lib/python2.6/doctest.py", line 1253, in __run

compileflags, 1) in test.globs

File "<doctest sympy.polys.ring_series.fun[4]>", line 1, in <module>

fun(p, _tan1, 'x', 4)

NameError: name 'fun' is not defined

===== tests finished: 2909 passed, 1 failed, 14 skipped, in 479.21 seconds =====

DO *NOT* COMMIT!
```

certik commented on sympy/polys/ring_series.py:
```diff
@@ -419,6 +418,61 @@ def rs_integrate(self, x):
+    of a multivariate series:
+
+      fun(p, tan, iv, prec)
+      compute _x.tan(iv, prec), then substitute _x with p
```
What is `_x` in this example?

certik commented on sympy/polys/ring_series.py:
```diff
@@ -419,6 +418,61 @@ def rs_integrate(self, x):
+    >>> R, x, y = ring('x, y', QQ)
+    >>> p = x + x*y + x**2*y + x**3*y**2
+    >>> fun(p, _tan1, 'x', 4)
+    1/3*x**3*y**3 + 2*x**3*y**2 + x**3*y + 1/3*x**3 + x**2*y + x*y + x
```
Reading this docstring and looking at examples, I still don't understand what the function `fun` is doing. Is this example taking some multivariate series `p`, and then calculating `tan(p)` by expanding into 4 terms of the taylor series for `tan`?

shivamvats replied:
I'm sorry, I will update it.
`fun` is used only for calculating multivariate series expansion. It calculates the series for a dummy variable `_x` and replaces `_x` with the intended variable `p`.

certik commented on sympy/polys/ring_series.py:
```diff
@@ -490,6 +544,111 @@ def rs_exp(p, x, prec):
+        p1 = rs_square(p, x, prec) + ring(1)
+        p1 = rs_series_inversion(p1, x, prec - 1)
+        p1 = rs_mul(dp, p1, x, prec - 1)
+        return rs_integrate(p1, x)
```
What are these 4 lines doing? Why are you differentiating and then integrating?

shivamvats replied:
Let's say I want to calculate `atan(p)`
`diff(atan(p)) = 1/(1+p**2) * diff(p)`
I then calculate the series expansion of RHS and integrate to get `atan(p)` back.
The same method is used for expansion of `log(x)`. This method is quite fast as polynomial differentiation and integration are trivial.

certik replied:
Ah I see. Clever.

certik commented:
It looks good, thanks! I left some comments. Also, don't forget to write regular tests.

shivamvats committed:
Fixed doctest of fun

shivamvats committed:
Added tests

shivamvats committed:
Added cot, cos, cos_sin, LambertW, nth_root, mul_xin and hyperbolic functions

shivamvats committed:
Tests added for cos, cos_sin and cot

shivamvats committed:
Added rs_diff and tests for atanh, sinh, cosh, tanh

shivamvats committed:
Fixed doctests

shivamvats committed:
Removed nth_root, asin and cot

These functions output Laurent or Puiseux series, which is not possible
currently. Will add then in a separate PR.

shivamvats committed:
Improved docstrings and added comments for better readability.

shivamvats commented:
@certik @thilinarmtb Could you please review the changes?

certik commented on sympy/polys/ring_series.py:
```diff
@@ -490,9 +622,344 @@ def rs_exp(p, x, prec):
+    >>> from sympy.polys.ring_series import rs_atan
+    >>> R, x = ring('x', QQ)
+    >>> rs_atan(x, x, 8)
+    -1/7*x**7 + 1/5*x**5 - 1/3*x**3 + x
```
You should add a doctest here showing that this function actually works when `p` is a series.

certik commented on sympy/polys/ring_series.py:
```diff
@@ -490,9 +622,344 @@ def rs_exp(p, x, prec):
+
+def _tan1(p, x, prec):
+    """
+    Helper function of ``rs_tan``
```
I saw you imported this in one of the doctests above. What exactly is this function doing? You should improve the docstring.

shivamvats replied:
@certik This function uses Newton's method to approximate `tan`. Say I want to calculate tan(p), I write it as `x = tan(p)` (assuming x to be the series expansion)  => `p  - atan(x) = 0`. Now I use Newton iterations to solve for x. I had assumed that this will work only for univariate p, but it seems to work even for multivariate p.
```
In [12]: (tan(a + a*b).series(a, 0, 5)).expand()
Out[12]: a + a*b + a**3/3 + a**3*b + a**3*b**2 + a**3*b**3/3 + O(a**5)
In [13]: _tan1(x + x*y, x, 5)
Out[13]: 1/3*x**3*y**3 + x**3*y**2 + x**3*y + 1/3*x**3 + x*y + x
```
If so, I think we can do away with the call to `fun`.

certik commented:
I think this looks very good, the tests seem thorough, +1 to merge if tests pass.

Are there other parts of #609 that are still not ported?

thilinarmtb commented:
I left some minor comments. +1 from me as well.

shivamvats committed:
Added better examples to docstrings

shivamvats committed:
Improved docstrings and edited atanh and atan

Improved docstrings of tan and tanh by explaining algo used.
Separated formula based evaluation of atan and atanh from main functions.

shivamvats commented:
@certik @thilinarmtb

shivamvats commented on sympy/polys/ring_series.py:
```diff
@@ -490,9 +622,363 @@ def rs_exp(p, x, prec):
+    p1 = rs_mul(dp, p1, x, prec - 1)
+    return rs_integrate(p1, x)
+
+def _tan1(p, x, prec):
```
@certik My previous comment about `_tan1` is not visible due to new commits. I am pasting it again:
This function uses Newton's method to approximate tan. Say I want to calculate tan(p), I write it as x = tan(p) (assuming x to be the series expansion) => p - atan(x) = 0. Now I use Newton iterations to solve for x. I had assumed that this will work only for univariate p, but it seems to work even for multivariate p.
```
In [12]: (tan(a + a*b).series(a, 0, 5)).expand()
Out[12]: a + a*b + a**3/3 + a**3*b + a**3*b**2 + a**3*b**3/3 + O(a**5)
In [13]: _tan1(x + x*y, x, 5)
Out[13]: 1/3*x**3*y**3 + x**3*y**2 + x**3*y + 1/3*x**3 + x*y + x
```
If so, I think we can do away with the call to fun.

certik replied:
I saw it, I was just thinking about it, whether this is the most efficient way, or whether you want to just use the regular series for `atan(x) = x - x**3/3 + x**5/5 - x**7/7 + x**9/9 + ...` and plug a series for x. How do you know that your algorithm is converged?

shivamvats replied:
I made an attempt to compare the two methods and it seems that the closed form method should be faster for atan and atanh (as they have simpler formulas). However I couldn't conclude anything for tan and tanh.

The algorithm should converge according to this http://www.mathstat.concordia.ca/faculty/rhall/mc/arctan.pdf

certik replied:
The link you posted shows the implementation of `rs_atan` and I agree that it should work. However, I am questioning the implementation of `_tan1`.

shivamvats replied:
That implementation is originally by Mario. I found a reference to it in http://www.loria.fr/~zimmerma/mca/mca-cup-0.5.9.pdf, page 130. 

Do you want me to replace it with the formula based method?

certik replied:
Paul Zimmermann is one of the best in the field, but I thought that book is about arbitrary precision arithmetic, not symbolic series like we have here.

Leave it as it is for now. We should make double sure it produces correct results (hopefully it does) and then benchmark which method is faster. We can do that later. It is well documented now what it does, so let's port the rest of Mario's work first, then worry about this.

thilinarmtb commented:
Looks good. Please address my previous comments as well.

shivamvats commented:
@thilinarmtb I am sorry I can only see your comments for `nth_root` which I had removed in a later commit. Did I miss some comments?

certik commented:
I am fine with this, unless @thilinarmtb has some comments that weren't addressed.

thilinarmtb commented:
Looks like you have corrected them. Sorry for the mistake. +1 from me as well.

certik commented:
Very well, let's merge it then. Thanks @shivamvats!

certik commented:
@pernici, you might be interested in this PR, that we just merged. @shivamvats ported a lot of your code from #609.

toolforger commented:
Am 08.06.2015 um 15:58 schrieb Shivam Vats:
> @thilinarmtb I am sorry I can only see your comments for `nth_root` which I had removed in a later commit. Did I miss some comments?

You can click on the "thilinarmtb commented on an outdated diff" link.
